src/hotspot/share/opto/vmaskloop.cpp

```diff
@@ -0,0 +1,975 @@
+        Node* out = node->fast_out(i);
+        if (!in_body(out)) {
+          trace_msg(node, "Node has out-of-loop user found");
+          return false;
+        }
+      }
+    }
+  }
+  // 4) Bail out if the loop body has extra node
+  for (int idx = 0; idx < n_nodes; idx++) {
+    Node* node = _body_nodes.at(idx);
+    if (!tracked.test(idx) && !in_core(node) && !is_memory_phi(node)) {
+      trace_msg(node, "Found extra loop node in loop body");
+      return false;
+    }
+  }
+  return true;
+}
+
+// Try creating a vector mask with the smallest vector element size
+const TypeVectMask* VectorMaskedLoop::create_vector_mask_type() {
+  BasicType vmask_bt = size_to_basic_type(_size_stats.smallest_size());
+  int vlen = Matcher::max_vector_size(vmask_bt);
+  if (!Matcher::match_rule_supported_vector(Op_LoopVectorMask, vlen, vmask_bt)) {
+    // Unable to create vector mask with the vlen & bt on this platform
+    return nullptr;
+  }
+  return (TypeVectMask*) TypeVect::makemask(vmask_bt, vlen);
+}
+
+// This checks if memory access node is our supported pattern
+bool VectorMaskedLoop::supported_mem_access(MemNode* mem) {
+  // First do a quick check by searching existing SWPointer(s)
+  for (int idx = 0; idx < _swptrs.length(); idx++) {
+    if (_swptrs.at(idx)->mem() == mem) {
+      return true;
+    }
+  }
+  // If not found, try creating a new SWPointer and insert it
+  SWPointer* ptr = mem_access_to_swpointer(mem);
+  if (ptr != nullptr) {
+    _swptrs.push(ptr);
+    return true;
+  }
+  return false;
+}
+
+// This tries creating an SWPointer object associated to the memory access.
+// Return nullptr if it fails or the SWPointer is not valid.
+SWPointer* VectorMaskedLoop::mem_access_to_swpointer(MemNode* mem) {
+  // Should access memory of a Java primitive value
+  BasicType mem_type = mem->memory_type();
+  if (!is_java_primitive(mem_type)) {
+    return nullptr;
+  }
+  // addp: memory address for loading/storing an array element. It should be an
+  // AddP node operating on an array of specific type
+  Node* addp = mem->in(MemNode::Address);
+  if (!addp->is_AddP() || !operates_on_array_of_type(addp, mem_type)) {
+    return nullptr;
+  }
+  // Create a Node_Stack for SWPointer's initial stack
+  Node_Stack* nstack = new Node_Stack(_arena, 5);
+  nstack->push(addp, 0);
+  // addp2: another possible AddP node for array element addressing. It should
+  // operate on the same memory type and have the same base with previous AddP.
+  Node* addp2 = addp->in(AddPNode::Address);
+  if (addp2->is_AddP()) {
+    if (!operates_on_array_of_type(addp2, mem_type) ||
+        addp->in(AddPNode::Base) != addp2->in(AddPNode::Base)) {
+      return nullptr;
+    }
+    nstack->push(addp2, 1);
+  }
+
+  // Check supported memory access via SWPointer. It's not supported if
+  //  1) The constructed SWPointer is invalid
+  //  2) Address is growing down (index scale * loop stride < 0)
```

Comment:
Is that a limitation that could be removed in the future?

Reply:
Yes, at least on SVE2. For growing up memory accesses, we generate vector masks that indicate active lanes at lower parts of a vector. But it's opposite for growing down memory accesses where active lanes are at higher parts of a vector. Only SVE2 of AArch64 can generate vector masks in this way, current SVE(1) can not. I'm not sure whether x86 AVX-512 has the similar ability.

Reply:
There must surely be some way. The only question is what is the cheapest way to do it, ie with the fewest number of instructions.